src/process/io/StreamSource.cpp

```diff
@@ -0,0 +1,268 @@
+}
+
+bool StreamSource::exists() {
+    return true;
+}
+
+/**
+ * What is the proper size for a stream? int64_t max?
+ * @return
+ */
+int64_t StreamSource::size() {
+    return -1;
+}
+
+int64_t StreamSource::getRewindBufferSize() {
+    int total = 0;
+    for (auto length : rewindBufferEntryLength) {
+        total += length;
+    }
+    return total;
+}
+
+char *StreamSource::joinRewindBuffer() {
+    char *buf = new char[getRewindBufferSize()];
+    int bufPos = 0;
+    for (u_int64_t i = 0; i < rewindBuffer.size(); i++) {
+        memcpy(buf + bufPos, rewindBuffer[i], static_cast<size_t>(rewindBufferEntryLength[i]));
+        bufPos += rewindBufferEntryLength[i];
+    }
+    return buf;
+}
+
+bool StreamSource::isFile() { return false; }
+
+bool StreamSource::isStream() { return true; }
+
+/**
+ * Reads numberOfBytes into the targetBuffer.
+ * @param targetBuffer targetBuffer is either a valid Byte array OR nullptr! In case of nullptr, the read bytes are
+ *                     still stored in the internal rewind buffer. Can e.g. be used by skip.
+ * @param numberOfBytes
+ * @return
+ */
+int64_t StreamSource::read(Bytef *targetBuffer, int numberOfBytes) {
+    if (rewoundBytes > 0) {
+        return readFromBuffer(targetBuffer, numberOfBytes);
+    } else {
+        return readFromStream(targetBuffer, numberOfBytes);
+    }
+}
+
+int64_t StreamSource::readFromBuffer(Bytef *targetBuffer, int numberOfBytes) {
+    char *buf = joinRewindBuffer();
+    int64_t bufLen = getRewindBufferSize();
+
+    int64_t copiedBytes = numberOfBytes;
+    if (numberOfBytes > bufLen) {
+        // Cut it, copy less!
+        copiedBytes = bufLen;
+    }
+
+    if (numberOfBytes > rewoundBytes) {
+        copiedBytes = rewoundBytes;
+    }
+
+    // Copy copiedBytes of data from the right side!
+    if (targetBuffer != nullptr)
+        memcpy(targetBuffer, buf + bufLen - rewoundBytes, copiedBytes); // TODO
+    rewoundBytes -= copiedBytes;
+    currentPosition += copiedBytes;
+
+    delete[] buf;
+
+    return copiedBytes;
+}
+
+int64_t StreamSource::readFromStream(Bytef *targetBuffer, int numberOfBytes) {
+    char *readBuffer = new char[numberOfBytes]{0};
+    inputStream->sync();
+
+    inputStream->read(readBuffer, numberOfBytes);
+    streamsize amountRead = inputStream->gcount();
+
+    if (amountRead > 0) {
+        if (targetBuffer != nullptr)
+            memcpy(targetBuffer, readBuffer, amountRead);// TODO
+        rewindBuffer.emplace_back(readBuffer);
+        rewindBufferEntryLength.emplace_back(amountRead);
+        if (rewindBuffer.size() > maxSegmentsInBuffer) {
+            char *front = rewindBuffer.front();
+            rewindBuffer.pop_front();
+            rewindBufferEntryLength.pop_front();
+            delete[] front;
+        }
+        currentPosition += amountRead;          // Advance current position marker
+    } else {
+        delete[] readBuffer;
+    }
+    return amountRead;
+}
+
+int StreamSource::readChar() {
```

Comment:
Could you document, that a negative return value means an error (right?). Also this is a typical C idiom and everything but expressive? Here you could use something like "Optional".

Reply:
See #44, it is described there.